internal/blocktx/processor.go

```diff
@@ -481,38 +401,69 @@ func (ph *Processor) processBlock(msg *p2p.BlockMessage) {
 	timeStart := time.Now()
 
 	blockHash := msg.Header.BlockHash()
-
 	previousBlockHash := msg.Header.PrevBlock
-
 	merkleRoot := msg.Header.MerkleRoot
 
-	blockId, err := ph.insertBlock(ctx, &blockHash, &merkleRoot, &previousBlockHash, msg.Height)
+	prevBlock, err := p.getPrevBlock(ctx, &previousBlockHash, &blockHash)
+	if err != nil {
+		p.logger.Error("unable to get previous block from db", slog.String("hash", blockHash.String()), slog.Uint64("height", msg.Height), slog.String("prevHash", previousBlockHash.String()), slog.String("err", err.Error()))
+		return
+	}
+
+	incomingBlock := createBlock(msg, prevBlock)
+
+	competing, err := p.competingChainsExist(ctx, incomingBlock)
+	if err != nil {
+		p.logger.Error("unable to check for competing chains", slog.String("hash", blockHash.String()), slog.Uint64("height", msg.Height), slog.String("err", err.Error()))
+		return
+	}
+
+	if competing {
+		p.logger.Info("Competing blocks found", slog.String("incoming block hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height))
+
+		greatestChainwork, err := p.hasGreatestChainwork(ctx, incomingBlock)
+		if err != nil {
+			p.logger.Error("unable to get the chain tip to verify chainwork", slog.String("hash", blockHash.String()), slog.Uint64("height", incomingBlock.Height), slog.String("err", err.Error()))
+			return
+		}
+
+		if greatestChainwork {
+			// TODO: perform reorg - next ticket
+			incomingBlock.Status = blocktx_api.Status_LONGEST
+		} else {
+			incomingBlock.Status = blocktx_api.Status_STALE
+		}
```

Comment:
nit:

`else` could be elimintated by 

```
incomingBlock.Status = blocktx_api.Status_STALE
if greatestChainwork {
  // TODO: perform reorg - next ticket
  incomingBlock.Status = blocktx_api.Status_LONGEST
} 
```

Reply:
I don't think it can be eliminated, let me explain.

In this implementation, the incoming block first gets the status of the previous block. So if the previous block was `LONGEST` the incoming block will initially also be `LONGEST`. This is done this way, because this will be the most likely and most common scenario for all incoming blocks (assuming that forks are rather rare).

Only then we check for competing blocks at the height of this incoming block and mark the block accordingly (STALE or LONGEST) based on chainwork.

So, the first competing block that will come at any height, will have the `LONGEST` block as it's previous block and will initially be also marked as `LONGEST`. Then we will see that this block is actually a competing block, so if it has a lower chainwork, we have to mark is as `STALE`.

This all happens before any write to the database, and with all possible edge cases I came up with, this is the easiest and most efficient way to do this. Hope I explained it well :)